src/drivers/intel/cavs/mn.c

```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Janusz Jankowski <janusz.jankowski@linux.intel.com>
+
+#include <sof/drivers/mn.h>
+#include <sof/drivers/ssp.h>
+#include <sof/lib/shim.h>
+#include <sof/math/numbers.h>
+#include <sof/trace/trace.h>
+
+/* tracing */
+#define trace_mn(__e, ...) \
+	trace_event(TRACE_CLASS_MN, __e, ##__VA_ARGS__)
+#define trace_mn_error(__e, ...) \
+	trace_error(TRACE_CLASS_MN, __e, ##__VA_ARGS__)
+#define tracev_mn(__e, ...) \
+	tracev_event(TRACE_CLASS_MN, __e, ##__VA_ARGS__)
+
+int mn_set_mclk(uint16_t mclk_id, uint32_t mclk_rate)
```

Comment:
this is a bit misleading. Here you don't actually support all cases of M/N, you assume an integer relationship with the ssp-freq.

If this was a true M/N support for the MCLK, then we'd need to allow for M different from 1, similar to the bclock?

Reply:
But we don't have M in M/N for MCLK, we just have mdiv, for me it's also weird but it's how hardware is implemented. The same goes for that XTAL bypassing, also it ignores M/N but is managed by M/N part of hardware and is in the same registers...

Reply:
ah, yes, but in that case it'd be useful to state it, I thought it was a miss when it was a feature

Reply:
Added that info to doxygen of this function - if function is globally visible then doxygen comments are in header file (mn.h)